Document Services for Extensions (#20376)
directus/directus: docs/extensions/services/authoring-database.md

---
description: Learn about the Services for authoring the database in Directus and how to utilize them when building extensions.
contributors: Esther Agbaje
---

# Authoring the Database

> Directus provides services such as `CollectionsService`, `FieldsService` and `RelationsService` to programmatically
> configure and modify the database schema.

Reviewer: ```suggestion
Use the `CollectionsService`, `FieldsService` and `RelationsService` to
configure and modify the data model.
```

## CollectionsService

The `CollectionsService` is responsible for managing and manipulating data in collections. Utilize it to create CRUD

> endpoints to collections.

Reviewer: Create CRUD is like Pin Number. 'Perform' might be better.

    const { collection, field } = req.params;
    const service = new FieldsService({ schema });
    await service.deleteField(collection, field);
    res.json();
  });
```

## RelationsService

The `RelationsService` allows you to manage relationships and references between items in Directus. Utilize it to create
endpoints that perform CRUD operations on relations.

### Create Relation

```js
router.post('/relations', async (req, res) => {
    const schema = await getSchema();

    const service = new RelationsService({ schema });

    const data = await service.createOne(req.body);

    const record = await service.readOne(data);
    res.locals['payload'] = { data: record || null };

    res.json(record);
  });
```

### Get Relation

```js
router.get('/relations/:collection/:field', async (req, res) => {
    const schema = await getSchema();

    const service = new RelationsService({ schema });
    const { collection, field } = req.params;

    const data = await service.readOne(collection, field);

    res.json(data);
  });
```

::: tip Explore Services In-depth

Check out the full list [here](https://github.com/directus/directus/blob/main/api/src/services).

> :::

Reviewer: Please apply the same changes to this file as the ItemsService page: 

1. Rename based on nav suggestions
2. Add a generic 'this is how you initialize' example before the first method. in each section. 
3. Fix up indentation. 
4. End of file make sure link contains explainer.
